src/models/OAuthCode.ts

```diff
@@ -0,0 +1,174 @@
+  cleanup(): void {
+    this.authorizationCode = this.authorizationCode || this.generateAuthorizationCode();
+    this.expiresAt = this.expiresAt || this.generateExpiryDate();
+    this.redirectUri = this.redirectUri || 'http://localhost:3000';
+    this.scope = this.scope || ['read'];
+  }
+
+  // Generate the token TTL
+  generateExpiryDate(): Date {
+    const currentDate = new Date();
+    return new Date(currentDate.getTime() + oauthConfig.authorizationCodeLifetime * 1000);
+  }
+
+  // Generate a new unique authorization code
+  generateAuthorizationCode(): string {
+    return uuidv4();
+  }
+
+  // Ensure that the authorization code has not expired
+  async validate(): Promise<boolean> {
+    if (this.authorizationCode) {
+      // Make sure it still exists in the database! and the payload matches (not tampered with)
+      const existing = await OAuthCode.findOne(this.authorizationCode);
+      if (existing && existing === this) {
+        // Make sure it is not expired
+        const currentDate = new Date();
+        return this.expiresAt > currentDate;
+      }
+      return false;
+    }
+    return false;
+  }
+
+  // Save the authorization code to the DB
+  async save(): Promise<boolean> {
+    this.cleanup();
+
+    const sql = `
+      INSERT INTO oauthCodes
+        (code, expiresAt, redirectUri, scope, userId, clientId, codeChallenge, codeChallengeMethod)
+      VALUES (?, ?, ?)
+    `;
+    const vals = [
+      this.authorizationCode,
+      this.expiresAt.toISOString(),
+      this.redirectUri,
+      stringToArray(this.scope).join(', '),
+      this.client.id.toString(),
+      this.user.id.toString(),
+      this.codeChallenge,
+      this.codeChallengeMethod,
+    ];
+
+    try {
+      const [result] = await mysql.query(sql, vals);
+      if (result[0]) {
+        console.log(`Authorization code created: User: ${this.user.id}, Code: ${result.insertId}`);
+        return true;
+      }
+      return false;
+    } catch(err) {
+      console.log('Error creating OAuthCode: ', err)
+      throw err;
+    }
+  }
+
+  // Set the authorization code expiry to now
+  async revoke(): Promise<boolean> {
+    const currentDate = new Date();
+    const sql = `UPDATE oauthCodes SET expiresAt = ?, modified = ? WHERE code = ?`;
+    const vals = [currentDate.toISOString(), currentDate.toISOString(), this.authorizationCode];
+    try {
+      const [result] = await mysql.query(sql, vals);
+      // TODO: Fix this Type issue, we should able to define one here
+      console.log(`Authorization code was revoked: ${this.authorizationCode}`);
+      return true;
+    } catch (err) {
+      console.log('Error revoking OAuthCode: ', err)
+      throw err;
+    }
+  }
+
+  // Convert the DB fields to properties of this call (except client and user)
+  static _SqlFieldsToProperties(row) {
+    const scopeArray = typeof row.scope === 'string' ? row.scope.split(' ') : row.scope;
+    return {
+      authorzationCode: row.code,
```

Comment:
L166: Typo - `authorzationCode` should be `authorizationCode`

Reply:
good catch